cmd/patch.go

```
@@ -0,0 +1,198 @@
+		return errors.Errorf("cannot patch non-exists cluster %s", clusterName)
+	}
+
+	if exist, err := utils.FileExist(packagePath); err != nil {
+		return err
+	} else if !exist {
+		return errors.New("specified package not exists")
+	}
+
+	metadata, err := meta.ClusterMetadata(clusterName)
+	if err != nil {
+		return err
+	}
+
+	insts, err := instancesToPatch(metadata, options)
+	if err != nil {
+		return err
+	}
+	if err := checkPackage(clusterName, insts[0].Role(), packagePath); err != nil {
+		return err
+	}
+
+	var replacePackageTasks []task.Task
+	for _, inst := range insts {
+		deployDir := clusterutil.Abs(metadata.User, inst.DeployDir())
+		tb := task.NewBuilder()
+		tb.BackupComponent(inst.ComponentName(), metadata.Version, inst.GetHost(), deployDir).
+			InstallPackage(packagePath, inst.GetHost(), deployDir)
+		replacePackageTasks = append(replacePackageTasks, tb.Build())
+	}
+
+	t := task.NewBuilder().
+		SSHKeySet(
+			meta.ClusterPath(clusterName, "ssh", "id_rsa"),
+			meta.ClusterPath(clusterName, "ssh", "id_rsa.pub")).
+		ClusterSSH(metadata.Topology, metadata.User, sshTimeout).
+		Parallel(replacePackageTasks...).
+		ClusterOperate(metadata.Topology, operator.UpgradeOperation, options).
+		Build()
+
+	if err := t.Execute(task.NewContext()); err != nil {
+		if errorx.Cast(err) != nil {
+			// FIXME: Map possible task errors and give suggestions.
+			return err
+		}
+		return errors.Trace(err)
+	}
+
+	if overwrite {
+		if err := overwritePatch(clusterName, insts[0].Role(), packagePath); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func instancesToPatch(metadata *meta.ClusterMeta, options operator.Options) ([]meta.Instance, error) {
+	roleFilter := set.NewStringSet(options.Roles...)
+	nodeFilter := set.NewStringSet(options.Nodes...)
+	components := metadata.Topology.ComponentsByStartOrder()
+	components = operator.FilterComponent(components, roleFilter)
+
+	instances := []meta.Instance{}
+	for _, com := range components {
```

I think we should raise an error if the instances contain two different kinds of components.